pkg/server/sandbox_portforward_windows.go

```diff
@@ -19,19 +19,68 @@ limitations under the License.
 package server
 
 import (
+	"bytes"
+	"fmt"
 	"io"
 
-	"github.com/containerd/containerd/errdefs"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
+	"k8s.io/client-go/tools/remotecommand"
 	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/util/ioutils"
+	"k8s.io/utils/exec"
+
+	ctrdutil "github.com/containerd/cri/pkg/containerd/util"
+	sandboxstore "github.com/containerd/cri/pkg/store/sandbox"
 )
 
 // PortForward prepares a streaming endpoint to forward ports from a PodSandbox, and returns the address.
 // TODO(windows): Implement this for windows.
 func (c *criService) PortForward(ctx context.Context, r *runtime.PortForwardRequest) (*runtime.PortForwardResponse, error) {
-	return nil, errdefs.ErrNotImplemented
+	sandbox, err := c.sandboxStore.Get(r.GetPodSandboxId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find sandbox %q", r.GetPodSandboxId())
+	}
+
+	if sandbox.Status.Get().State != sandboxstore.StateReady {
+		return nil, errors.New("sandbox container is not running")
+	}
+
+	return c.streamServer.GetPortForward(r)
 }
 
 func (c *criService) portForward(ctx context.Context, id string, port int32, stream io.ReadWriter) error {
-	return errdefs.ErrNotImplemented
+	_, err := c.sandboxStore.Get(id)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find sandbox %q in store", id)
+	}
+	stderr := new(bytes.Buffer)
+	err = c.ExecInSandbox(id, []string{"wincat.exe", "localhost", fmt.Sprint(port)}, stream, ioutils.WriteCloserWrapper(stream), ioutils.WriteCloserWrapper(stderr), false, nil)
+	if err != nil {
+		return fmt.Errorf("%v: %s", err, stderr.String())
+	}
+
+	return nil
+}
+
+func (c *criService) ExecInSandbox(sandboxID string, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser,
```

Comment:
Keep the function private, we don't need to export it yet.

Reply:
We can remove `tty` and `resize` from the function signature if we don't need them any time soon.

Reply:
Pass in the context and use it instead of creating a new one.

Reply:
Done. The context needs to have a namespace so I created it using the `ctrdutil.NamespacedContext()`.

Reply:
The context passed to portforward should be namespaced, if not we should fix it. :) https://github.com/containerd/cri/blob/master/pkg/server/streaming.go#L159

Reply:
Yeah I'm getting this error ` failed to get container spec: namespace is required: failed precondition` if directly passing the `ctx`. I'll looking into it.